############################################################
##### 02. bulk RNA-seq Processing & Differential Gene Expression (DGE)
##### April 2025, Hyojin Kang
############################################################
library("Biobase")
library("DESeq2")
library("ggplot2")
library("tximport")
library("readr")

############################################################
##### Pipeline #####
### (1) Load count data
### (2) Batch correction
### (3) Gene Filtering
### (4) Normalization
### (5) Differential Expressed Genes (DEGs)
############################################################
#####(1) Load count data
### Differential gene expression (DGE) analysis was performed independently for each mouse strain and sex
wkdir="/path/to/my/directory/"
GENE <- "ADNP"
TAG <- "ADNP_Male"
setwd(paste(wkdir, TAG, sep=""))

sampleList <- list.files(paste(wkdir, TAG, "/salmon", sep=""))
sampleFiles <- file.path(paste(wkdir, TAG, sep=""), "salmon",sampleList, "quant.genes.sf")


genotype <- c(rep("HT",15), rep("WT",15))
drug <- c(rep("fluoxetine",5), rep("vehicle",5), rep("fluoxetine",5), rep("vehicle",5))
group <- c(rep("HfM",5), rep("HLM",5), rep("HvM",5), rep("WfM",5), rep("WLM",5), rep("WvM",5))
batch <- factor(c(rep("batch1",5), rep("batch2",5),rep("batch1",5),rep("batch2",5),rep("batch1",5),rep("batch2",5)))

samples <- data.frame(sampleName = sampleList, fileName = sampleFiles, condition=group, batch=batch)

### Load Ensembl Annotation #####
library(biomaRt)
ensembl_mm10_mart <- useMart(biomart="ENSEMBL_MART_ENSEMBL", dataset="mmusculus_gene_ensembl")
tx2gene <- getBM(attributes=c("ensembl_gene_id", "ensembl_gene_id"), mart=ensembl_mm10_mart)
#httr::set_config(httr::config(ssl_verifypeer = FALSE))
annotation <- getBM(attributes=c("ensembl_gene_id", "external_gene_name","entrezgene_accession", "chromosome_name", "start_position", "end_position", "description"), mart=ensembl_mm10_mart)
names(tx2gene) <- c("TXNAME", "GENEID")

### import salmon count data using tximport function
txi <- tximport(sampleFiles, type="salmon", tx2gene=tx2gene, ignoreTxVersion=TRUE)

############################################################
##### (2) Batch correction #####
### Bulk RNA-seq data were generated across eight separate batches. 
### To correct for potential batch effects, the ComBat_seq function from the R/Bioconductor sva (v3.42.0) package was applied to the raw count data. 
### ComBat-seq takes as input an untransformed count matrix along with known batch variables. 
### Batch effects were modeled using a negative binomial regression, 
### and adjusted expression values were obtained by mapping the original data to an expected distribution under the assumption of no batch effects.
### batch-specific dispersion parameters are estimated, allowing for flexible modeling of gene expression variability across batches. 
### The adjusted data generated by ComBat-seq retain the integer nature of the original counts, 
### thereby ensuring compatibility with downstream differential expression analysis tools such as DESeq2, which require untransformed count data.

ddsTxi <- DESeqDataSetFromTximport(txi, colData = samples, design = ~ condition)
ddsCount <- counts(ddsTxi)
saveRDS(ddsCount, paste(DIR, TAG, "/rdsDdsCount.RData", sep=""))
combatCount <- ComBat_seq(counts = ddsCount, batch = batch)
ddsCombat <- DESeqDataSetFromMatrix(countData=combatCount, colData = samples, design = ~condition)

############################################################
##### (3) Gene Filtering #####
### Raw count data were initially prefiltered to remove lowly expressed genes by excluding those with total expression levels less than five times the number of samples
### by excluding those with total expression levels less than five times the number of samples
### Keep genes with raw read counts > # of samples * 5
keep <- rowSums(counts(ddsCombat)) >= length(sampleFiles)*5
ddsCombat <- ddsCombat[keep,]

############################################################
##### (4) Normalization #####
### The DESeq2 function automatically performs 1) regularization, 2)variance estimation, and 3)statistical modeling in a step-by-step manner.
dds <- DESeq(ddsCombat, parallel=TRUE)

### (step 1) Calculating Size Factor (Normalization)
### dds <- estimateSizeFactors(dds)

### (step 2) Dispersion Estimation
### dds <- estimateDispersions(dds)

### (step 3) Fitting a GLM model (Generalized Linear Model)
### dds <- nbinomWaldTest(dds)

############################################################
##### (5) Differential Expressed Genes (DEGs) #####
res <- results(dds, contrast=c("condition", "HvM", "WvM"), pAdjustMethod = "BH", alpha=0.05)
res <- res[!is.na(res$pvalue),]
write.table(as.data.frame(res), file=paste(GENE, "_deseq2_", TAG, "_all.txt", sep=""), sep="\t", quote = FALSE)

